File: my-folder/0380-insert-delete-getrandom-o1/solution.java
class RandomizedSet {

    java.util.Random rand = new java.util.Random();
    Map<Integer, Integer> map;
    List<Integer> list;

    public RandomizedSet() {
        map = new HashMap<>();
        list = new ArrayList<>();
    }
    
    public boolean insert(int val) {
        if (map.containsKey(val)) return false;
        map.put(val, list.size());
        list.add(val);
        return true;
    }
    
    public boolean remove(int val) {
        if (!map.containsKey(val)) return false;
        int location = map.get(val);
        if (location < list.size() - 1) {
            int last = list.get(list.size() - 1);
            list.set(location, last);
            map.put(last, location);
        }
        map.remove(val);
        list.remove(list.size() - 1);
        return true;
    }
    
    public int getRandom() {
        int random = rand.nextInt(list.size());
        return list.get(random);
    }
}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet obj = new RandomizedSet();
 * boolean param_1 = obj.insert(val);
 * boolean param_2 = obj.remove(val);
 * int param_3 = obj.getRandom();
 */

File: my-folder/0380-insert-delete-getrandom-o1/solution.py
class RandomizedSet:

    def __init__(self):
        self.arr = []
        self.hash_map = {}

    def insert(self, val: int) -> bool:
        if val not in self.hash_map:
            self.arr.append(val)
            self.hash_map[val] = len(self.arr) - 1
            return True
        return False

    def remove(self, val: int) -> bool:
        if val in self.hash_map:
            location = self.hash_map[val]
            if location < len(self.arr) - 1:
                last = self.arr[-1]
                self.hash_map[last], self.arr[location] = location, last
            self.arr.pop()
            self.hash_map.pop(val)
            return True
        return False

    def getRandom(self) -> int:
        loc = random.randint(0, len(self.arr) - 1)
        return self.arr[loc]


# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()
